[GHF] Better check for internal diffs

malfet opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #104344
* #104343

During revert, use title of "Meta Internal-Only Changes Check" to determine whether or not internal diff is associated with the PR. When PR is merged/closed, "Meta Internal-Only Changes Check" status is always success, but title message can differ:
- "There is no internal Diff connected, this can be merged now" means that there are no internal change associated with PR (or it was landed via GitHub First workflow)
- "The internal Diff has landed, this can be merged now" meaning that PR has associated internal DIFF, and OSS and internal reverts must happen in sync using internal tooling. (Or a revert PR can be authored in OSS)

Add regression test for https://github.com/pytorch/pytorch/pull/100652 that was originated from the internal diff, but was merged as OSS PR.

Fixes https://github.com/pytorch/pytorch/issues/104232

cc @albanD

malfet committed:
[GHF] Better check for internal diffs

During revert, use title of "Meta Internal-Only Changes Check" to
determine whether or not internal diff is associated with the PR.

Fixes https://github.com/pytorch/pytorch/issues/104232

[ghstack-poisoned]

malfet committed:
Update on "[GHF] Better check for internal diffs"


During revert, use title of "Meta Internal-Only Changes Check" to determine whether or not internal diff is associated with the PR.

Fixes https://github.com/pytorch/pytorch/issues/104232

cc albanD

[ghstack-poisoned]

malfet committed:
Update on "[GHF] Better check for internal diffs"


During revert, use title of "Meta Internal-Only Changes Check" to determine whether or not internal diff is associated with the PR.

Fixes https://github.com/pytorch/pytorch/issues/104232

cc albanD

[ghstack-poisoned]

bigfootjon commented on .github/scripts/trymerge.py:
```diff
@@ -977,14 +987,21 @@ def get_diff_revision(self) -> Optional[str]:
             return False
         return checks[checkrun_name].status != "SUCCESS"
 
+    def has_no_connected_diff(self) -> bool:
```
Returning a negative of a boolean value reduces readability, IMO replace this with `has_connected_diff`

malfet replied:
Sure, but is it safe to say, that `has_connect_diff()` is true if status is anything but "There is no internal Diff connected, this can be merged now"?

bigfootjon commented on .github/scripts/test_trymerge.py:
```diff
@@ -470,6 +464,16 @@ def commit_message(self, ref: str) -> str:
             lambda: validate_revert(repo, pr, comment_id=1372496233),
         )
 
+    def test_revert_codev_abandonned_diff_succeeds(self, *args: Any) -> None:
```
Typo: abandoned

bigfootjon reviewed (approved):
Please import this to fbsource to verify that the internal test suite will continue to work. I assume it will but this has been annoying to fix in the past

malfet committed:
Update on "[GHF] Better check for internal diffs"


During revert, use title of "Meta Internal-Only Changes Check" to determine whether or not internal diff is associated with the PR.

Fixes https://github.com/pytorch/pytorch/issues/104232

cc albanD

[ghstack-poisoned]

malfet commented:
> Please import this to fbsource to verify that the internal test suite will continue to work. I assume it will but this has been annoying to fix in the past

@bigfootjon I've not change any of the existing APIs nor added new dependencies, so imo it should be fine, but just to be on the safe side: https://www.internalfb.com/intern/sandcastle/job/18014399493129784/

malfet commented:
@pytorchbot merge -f "Lint and relevant internal tests are green"
